compiler/ballerina-tools-api/src/main/java/io/ballerina/tools/diagnostics/DefaultDiagnostic.java

```diff
@@ -0,0 +1,64 @@
+/*
+ *  Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package io.ballerina.tools.diagnostics;
+
+import java.text.MessageFormat;
+import java.util.List;
+
+/**
+ * An internal implementation of the {@code Diagnostic} class that is used by the {@code DiagnosticFactory}
+ * to create diagnostics.
+ *
+ * @since 2.0.0
+ */
+class DefaultDiagnostic extends Diagnostic {
+    private final DiagnosticInfo diagnosticInfo;
+    private final Location location;
+    private final List<DiagnosticProperty<?>> properties;
+    private final String message;
+
+    DefaultDiagnostic(DiagnosticInfo diagnosticInfo,
+                      Location location,
+                      List<DiagnosticProperty<?>> properties,
+                      Object[] args) {
+        this.diagnosticInfo = diagnosticInfo;
+        this.location = location;
+        this.properties = properties;
+        this.message = MessageFormat.format(diagnosticInfo.messageFormat(), args);
```

Comment:
This can cause IllegalArgumentException if args and message format do not match. This can cause BAD SAD.

Reply:
It seems like we haven't handled this in other places too. Maybe we can fix this separately.

Reply:
Oh okay.. I will update this with another PR soon

Reply:
I handled this by catching any compiler extension level error and re-throwing them as `ProjectException` to avoid BAD SADs. Just pushed the commit